File: iterableapi/src/test/java/com/iterable/iterableapi/TaskSchedulerTest.java
package com.iterable.iterableapi;

import com.iterable.iterableapi.unit.TestRunner;

import org.json.JSONObject;
import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;

import static org.mockito.ArgumentMatchers.any;
import static org.mockito.ArgumentMatchers.eq;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

@RunWith(TestRunner.class)
public class TaskSchedulerTest {
    private IterableTaskStorage mockTaskStorage;
    private IterableTaskRunner mockTaskRunner;
    private TaskScheduler taskScheduler;

    @Before
    public void setUp() throws Exception {
        mockTaskStorage = mock(IterableTaskStorage.class);
        mockTaskRunner = mock(IterableTaskRunner.class);
        taskScheduler = new TaskScheduler(mockTaskStorage, mockTaskRunner);
    }

    @Test
    public void testScheduleTaskCreatesTaskInStorage() throws Exception {
        IterableApiRequest request = new IterableApiRequest("apiKey", "api/test", new JSONObject(), "POST", null, null, null);
        taskScheduler.scheduleTask(request, null, null);
        verify(mockTaskStorage).createTask(eq("api/test"), eq(IterableTaskType.API), eq(request.toJSONObject().toString()));
    }

    @Test
    public void testSuccessCallbackIsCalledOnCompletion() throws Exception {
        IterableHelper.SuccessHandler successHandler = mock(IterableHelper.SuccessHandler.class);
        IterableApiRequest request = new IterableApiRequest("apiKey", "api/test", new JSONObject(), "POST", null, null, null);
        when(mockTaskStorage.createTask(any(String.class), any(IterableTaskType.class), any(String.class))).thenReturn("testTaskId");
        taskScheduler.scheduleTask(request, successHandler, null);
        taskScheduler.onTaskCompleted("testTaskId", IterableTaskRunner.TaskResult.SUCCESS, IterableApiResponse.success(200, "", new JSONObject()));
        verify(successHandler).onSuccess(any(JSONObject.class));
    }

    @Test
    public void testFailureCallbackIsCalledOnCompletion() throws Exception {
        IterableHelper.FailureHandler failureHandler = mock(IterableHelper.FailureHandler.class);
        IterableApiRequest request = new IterableApiRequest("apiKey", "api/test", new JSONObject(), "POST", null, null, null);
        when(mockTaskStorage.createTask(any(String.class), any(IterableTaskType.class), any(String.class))).thenReturn("testTaskId");
        taskScheduler.scheduleTask(request, null, failureHandler);
        taskScheduler.onTaskCompleted("testTaskId", IterableTaskRunner.TaskResult.FAILURE, IterableApiResponse.failure(400, "", new JSONObject(), "TestError"));
        verify(failureHandler).onFailure(eq("TestError"), any(JSONObject.class));
    }
}

File: iterableapi/src/test/java/com/iterable/iterableapi/TaskSchedulerTest.kt
package com.iterable.iterableapi

import com.iterable.iterableapi.unit.TestRunner
import org.json.JSONObject
import org.junit.Before
import org.junit.Test
import org.junit.runner.RunWith
import org.mockito.ArgumentMatchers.any
import org.mockito.ArgumentMatchers.eq
import org.mockito.Mockito.mock
import org.mockito.Mockito.verify
import org.mockito.Mockito.`when`

@RunWith(TestRunner::class)
class TaskSchedulerTest {
    private lateinit var taskStorage: IterableTaskStorage
    private lateinit var taskRunner: IterableTaskRunner
    private lateinit var requestDispatcher: IterableRequestDispatcher
    private lateinit var taskScheduler: TaskScheduler

    @Before
    fun setUp() {
        TaskScheduler.successCallbackMap.clear()
        TaskScheduler.failureCallbackMap.clear()
        taskStorage = mock(IterableTaskStorage::class.java)
        taskRunner = mock(IterableTaskRunner::class.java)
        requestDispatcher = mock(IterableRequestDispatcher::class.java)
        taskScheduler = TaskScheduler(
            taskStorage,
            taskRunner,
            requestDispatcher
        )
    }

    @Test
    fun `scheduling a request stores its serialized data`() {
        val request = request()

        taskScheduler.scheduleTask(request, null, null)

        verify(taskStorage).createTask(
            "api/test",
            IterableTaskType.API,
            request.toJSONObject().toString()
        )
    }

    @Test
    fun `storage failure dispatches the original request immediately`() {
        val success = mock(IterableHelper.SuccessHandler::class.java)
        val failure = mock(IterableHelper.FailureHandler::class.java)
        val request = request(success, failure)
        `when`(
            taskStorage.createTask(
                any(String::class.java),
                any(IterableTaskType::class.java),
                any(String::class.java)
            )
        ).thenReturn(null)

        taskScheduler.scheduleTask(request, success, failure)

        verify(requestDispatcher).execute(request)
    }

    @Test
    fun `successful stored request calls its client success callback`() {
        val success = mock(IterableHelper.SuccessHandler::class.java)
        val request = request(success, null)
        `when`(
            taskStorage.createTask(
                any(String::class.java),
                any(IterableTaskType::class.java),
                any(String::class.java)
            )
        ).thenReturn("task-id")
        taskScheduler.scheduleTask(request, success, null)

        val responseData = JSONObject()
        taskScheduler.onTaskCompleted(
            "task-id",
            IterableTaskRunner.TaskResult.SUCCESS,
            IterableApiResponse.success(200, "{}", responseData)
        )

        verify(success).onSuccess(responseData)
    }

    @Test
    fun `failed stored request calls its client failure callback`() {
        val failure = mock(IterableHelper.FailureHandler::class.java)
        val request = request(null, failure)
        `when`(
            taskStorage.createTask(
                any(String::class.java),
                any(IterableTaskType::class.java),
                any(String::class.java)
            )
        ).thenReturn("task-id")
        taskScheduler.scheduleTask(request, null, failure)

        val responseData = JSONObject()
        taskScheduler.onTaskCompleted(
            "task-id",
            IterableTaskRunner.TaskResult.FAILURE,
            IterableApiResponse.failure(
                400,
                """{"msg":"Bad request"}""",
                responseData,
                "Bad request"
            )
        )

        verify(failure).onFailure(eq("Bad request"), eq(responseData))
    }

    private fun request(
        success: IterableHelper.SuccessHandler? = null,
        failure: IterableHelper.FailureHandler? = null
    ): IterableApiRequest {
        return IterableApiRequest(
            "api-key",
            "api/test",
            JSONObject(),
            IterableApiRequest.POST,
            null,
            success,
            failure
        )
    }
}
